File: water/src/broker/IBrokerConnection.ts
import type { IBrokerMessageHeaders } from "./IBrokerMessageHeaders.js";


export type BrokerTopicListener = (key: string, value: string, headers: IBrokerMessageHeaders) => void | Promise<void>;

export interface IBrokerConnection {

    get clientId(): string;
    get clusterId(): string;

    start(): Promise<void>;
    destroy(): Promise<void>;

    on(topic: string, cb: BrokerTopicListener): void;
    off(topic: string, cb: BrokerTopicListener): void;

    send(
        topic: string,
        key: string,
        value: string,
        headers: IBrokerMessageHeaders,
    ): Promise<string>;

    createHeaders(
        targetClusters?: Set<string>,
        requestId?: string,
    ): IBrokerMessageHeaders;

}

File: latte/src/main/java/gg/beemo/latte/broker/IBrokerConnection.kt
package gg.beemo.latte.broker

fun interface TopicListener {
    suspend fun onMessage(key: String, value: String, headers: IBrokerMessageHeaders)
}

interface IBrokerConnection<Headers : IBrokerMessageHeaders> {

    val clientId: String
    val clusterId: String

    fun start()
    fun destroy()

    fun on(topic: String, cb: TopicListener)
    fun off(topic: String, cb: TopicListener)

    suspend fun send(
        topic: String,
        key: String,
        value: String,
        headers: Headers,
        blocking: Boolean = true,
    ): String

    fun createHeaders(targetClusters: Set<String>? = null, requestId: String? = null): IBrokerMessageHeaders

}
